src/CoreGraphics/CGEvent.cs

```diff
@@ -34,23 +34,70 @@ namespace CoreGraphics {
 	public sealed class CGEvent : NativeObject {
 		public delegate IntPtr CGEventTapCallback (IntPtr tapProxyEvent, CGEventType eventType, IntPtr eventRef, IntPtr userInfo);
 
+#if NET
+		class TapData {
+			public TapData (CGEventTapCallback cb, IntPtr userInfo)
+			{
+				Callback = cb;
+				UserInfo = userInfo;
+			}
+
+			public CGEventTapCallback Callback { get; private set; }
+			public IntPtr UserInfo { get; private set; }
+		}
+
+		[UnmanagedCallersOnly]
+		static IntPtr TapCallback (IntPtr tapProxyEvent, CGEventType eventType, IntPtr eventRef, IntPtr userInfo)
+		{
+			var gch = GCHandle.FromIntPtr (userInfo);
+			var tapData = (TapData)gch.Target!;
+			return tapData.Callback (tapProxyEvent, eventType, eventRef, tapData.UserInfo);
+		}
+		
+		[DllImport (Constants.ApplicationServicesCoreGraphicsLibrary)]
+		extern static unsafe IntPtr CGEventTapCreate (CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, delegate* unmanaged<IntPtr, CGEventType, IntPtr, IntPtr, IntPtr> cback, IntPtr data);
+#else
 		[DllImport (Constants.ApplicationServicesCoreGraphicsLibrary)]
 		extern static IntPtr CGEventTapCreate (CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, CGEventTapCallback cback, IntPtr data);
+#endif
 
 		public static CFMachPort? CreateTap (CGEventTapLocation location, CGEventTapPlacement place, CGEventTapOptions options, CGEventMask mask, CGEventTapCallback cback, IntPtr data)
 		{
+#if NET
+			IntPtr r;
+			unsafe {
+				var tapData = new TapData (cback, data);
```

Comment:
With regards to lifetime: it seems the `data` parameter has to stay alive until the tap is released, which we can assume is when the CFMachPort instance is collected.

Handling this can be accomplished by using a [CondtionalWeakTable](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-6.0) in this class:

```cs
public sealed class CGEvent : NativeObject {
	static ConditionalWeakTable<CFMachPort, TapData> tap_table;
	static object tap_lock = new object ();

	class TapData {
		public TapData (CGEventTapCallback cb, IntPtr userInfo)
		{
			Callback = cb;
			UserInfo = userInfo;
			handle = GCHandle.Alloc (this, GCHandleType.Weak);
		}
		~TapData ()
		{
			handle.Free ();
		}
		GCHandle handle;
		public CGEventTapCallback Callback { get; private set; }
		public IntPtr UserInfo { get; private set; }
		public IntPtr Handle { get { return GCHandle.ToIntPtr (handle); } }
	}

	public static CFMachPort? CreateTap (...)
	{
		var tapData = new TapData (cback, data);
		r = CGEventTapCreate (location, place, options, mask, &TapCallback, tapData.Handle);
		if (r == IntPtr.Zero)
			return null;
		var rv = new CFMachPort (r, true);
		lock (tap_lock) {
			if (tap_table is null)
				tap_table = new ConditionalWeakTable<CFMachPort, TapData> ();
			tap_table [rv, tapData];
		}
		return rv;
	}
}
```

And then the tap data will be collected once the CFMachPort is collected.

Note that this uses a weak GCHandle so that the GCHandle doesn't prevent the tap data from being collected.